src/basic/user-util.c

```diff
@@ -670,40 +670,32 @@ int reset_uid_gid(void) {
 }
 
 int take_etc_passwd_lock(const char *root) {
-
         struct flock flock = {
                 .l_type = F_WRLCK,
                 .l_whence = SEEK_SET,
                 .l_start = 0,
                 .l_len = 0,
         };
 
-        const char *path;
-        int fd, r;
-
         /* This is roughly the same as lckpwdf(), but not as awful. We don't want to use alarm() and signals,
          * hence we implement our own trivial version of this.
          *
          * Note that shadow-utils also takes per-database locks in addition to lckpwdf(). However, we don't,
          * given that they are redundant: they invoke lckpwdf() first and keep it during everything they do.
          * The per-database locks are awfully racy, and thus we just won't do them. */
 
-        if (root)
-                path = prefix_roota(root, ETC_PASSWD_LOCK_PATH);
-        else
-                path = ETC_PASSWD_LOCK_PATH;
+        _cleanup_free_ char *path = path_join(root, ETC_PASSWD_LOCK_PATH);
+        if (!path)
+                return log_oom_debug();
```

Comment:
you could go even further and use chase_symlinks_and_open()? in case /etc/ is a symlink? (would require CHASE_NONEXISTANT so that O_CREAT makes sense)

i mean, if you touch this, might be the best to just go in all the way, in particular as the code would be shorter then, even.

Reply:
Can we get the firstboot reset PR in which adds all the patterns that make this nice?

Reply:
Hmm, no, I don't think so. With my version, the behaviour is simple: it'll create directories, but any dangling symlinks will cause an error. I don't think we should follow dangling symlinks and create directories at the target: that's just too much magic and most likely a result of some screwup.